### PR TITLE
optimize Dict{Void}/Set{Void} (fix #20903)

### LINKED ISSUE #20903: Enhancement request: optimize Set{Void}
Ref: https://discourse.julialang.org/t/strange-behavior-with-void/2480

Would it be possible to special-case creating a set out of a large vector of Voids? This is admittedly a weird thing to ask, but it looks like it will be significant for some upcoming LightGraphs work. Right now, if I have a large vector of Voids, Julia knows not to allocate any memory for that vector, but once I try turning it into a Set, it appears that sizehint! is applied based on the number of elements, and this allocates a large amount of memory for a single-element Dict.

### PR DESCRIPTION
**Pull request by rfourquet**
A simple (simplistic?) fix for @sbromberger's problem (expressed in particular [here](https://github.com/JuliaLang/julia/issues/20903#issuecomment-284427298)) with the inefficiency of `Set(Vector{Void}(bignumber))`, using mainly ideas from https://github.com/JuliaLang/julia/issues/20903#issuecomment-284404696. This is maybe not a very general solution, but at least it can serve as a start.

### CODE REVIEW

File: base/set.jl
Comment (TotalVerb) [inline base/set.jl]: While we're at it, should this maybe dispatch on iteratorsize HasLength/HasSize instead of AbstractArray?

File: base/dict.jl
Comment (TotalVerb) [inline base/dict.jl]: For Bool, this is a substantial overestimate.

Comment (KristofferC) [inline base/dict.jl]: Why not just add a separate method  for Bool?

File: base/set.jl
Comment (rfourquet) [inline base/set.jl]: Yes good idea.

File: base/dict.jl
Comment (rfourquet) [inline base/dict.jl]: Ah of course, an oversight.

File: base/dict.jl
Comment (rfourquet) [inline base/dict.jl]: I feel uncomfortable with this one (in particular on 32 bit machines): `sizehint!` doesn't seem to make any garantee, but for `union!(s::Set, xs)`, could it be that `s` could contain more than `typemax(Int)` elements?

Comment (TotalVerb) [inline base/dict.jl]: I think this is safe. Such a set would probably malfunction badly with a `Dict` that has a negative underlying `count`...

**Comment by rfourquet** [comment]
Is someone opposed to this change? 
@JeffBezanson are you ok with the introduction here of `haslength` in base/generator.jl ?

**Comment by StefanKarpinski** [comment]
@JeffBezanson has been on vacation but is back and should be able to answer soon.

**Comment by KristofferC** [comment]
Is this good to go? @JeffBezanson

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Are we sure there's no slowdown in `union!` with this?

File: base/set.jl
Comment (tkelman) [inline base/set.jl]: space after comma

**Comment by rfourquet** [comment]
> Are we sure there's no slowdown in `union!` with this?

I didn't do thorough testing, but I didn't observe a slowdown by more than 5%.

File: base/set.jl
Comment (rfourquet) [inline base/set.jl]: OK, I added a space in few similar places in this file.

**Comment by rfourquet** [comment]
Rebased, will merge in a day or two if no objection.

**Comment by rfourquet** [comment]
1 travis failure is "time limit exceeded", appveyor failure related to libgit2

**Comment by musm** [comment]
should maybe add benchmarks to catch regressions?

**Comment by tkelman** [comment]
should have probably run nanosoldier here as well

**Comment by rfourquet** [comment]
> should maybe add benchmarks to catch regressions?

yes good idea, will do that

> should have probably run nanosoldier here as well

it's not possible on a merged PR?

**Comment by tkelman** [comment]
would have been best to do before merging

**Pull request merged**
